[FSDP][state_dict][2/N] Move state_dict related enums/dataclasses/states to state_dict_utils.py, api.py and init_state_dict()

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #88638
* #88637
* #88636
* #88635
* __->__ #88481
* #87900

**Motivation**:
Several Enums, Dataclasses and states defined in fully_sharded_data_paralle.py should be moved to a place where the composable FSDP can access. This PR does the move.

fegin committed:
[FSDP][state_dict][2/N] Move state_dict related enums/dataclasses/states to state_dict_utils.py and init_state_dict()

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][state_dict][2/N] Move state_dict related enums/dataclasses/states to state_dict_utils.py and init_state_dict()"




[ghstack-poisoned]

fegin committed:
Update on "[FSDP][state_dict][2/N] Move state_dict related enums/dataclasses/states to state_dict_utils.py and init_state_dict()"




[ghstack-poisoned]

awgu commented:
I am wondering if it is possible to / if we should move the public-facing state dict classes to `api.py`. I am pretty novice at figuring out these imports and public bindings though.

What would the import look like if a user wanted to get a `StateDictConfig` or `StateDictType`? In my understanding, we should not have users import from `torch.distributed.fsdp._state_dict_utils`. (That was why I moved classes like `BackwardPrefetch`, `ShardingStrategy`, etc. that the user interacts with to `api.py`. Classes like `TrainingState` which are only used internally are in `_common_utils.py`.)

fegin commented:
@awgu Users should not use `torch.distributed.fsdp._state_dict_utils` but should use `torch.distributed.fsdp`. Any public interfaces should be import in `torch.distributed.fsdp.__init__.py`. Then it won't matter where are those interfaces are defined internally. But I'm open to move to `api.py` if you prefer.

awgu commented:
> @awgu Users should not use `torch.distributed.fsdp._state_dict_utils` but should use `torch.distributed.fsdp`. Any public interfaces should be import in `torch.distributed.fsdp.__init__.py`. Then it won't matter where are those interfaces are defined internally. But I'm open to move to `api.py` if you prefer.

This makes sense. (No need to move to `api.py`.) I think everything should be fine since both `_composable/fully_shard.py` and `fsdp/fully_sharded_data_parallel.py` can import from `_state_dict_utils.py` without issue. Plus, we expose in `fsdp/__init__.py`.

rohan-varma reviewed (approved):
LG

fegin committed:
Update on "[FSDP][state_dict][2/N] Move state_dict related enums/dataclasses/states to state_dict_utils.py and init_state_dict()"




[ghstack-poisoned]

fegin committed:
Update on "[FSDP][state_dict][2/N] Move state_dict related enums/dataclasses/states to state_dict_utils.py, api.py and init_state_dict()"


**Motivation**:
Several Enums, Dataclasses and states defined in fully_sharded_data_paralle.py should be moved to a place where the composable FSDP can access. This PR does the move.



[ghstack-poisoned]

awgu reviewed (approved):
LGTM!

fegin commented on torch/distributed/fsdp/_state_dict_utils.py:
```diff
@@ -71,7 +100,10 @@ def _enter_full_param_ctx(
 
 def _exit_full_param_ctx(module) -> None:
     """A helper function to exit ``summon_full_params`` context."""
-    module._assert_state([TrainingState.SUMMON_FULL_PARAMS])
```
Using `TrainingState` is problematic for state_dict_utils as composable FSDP only has one training state. Using handle state makes more sense and will be done in the following PRs.

fegin commented:
@pytorchbot merge

fegin commented:
@pytorchbot merge -f "Failing test is not related."
